File: test/language/array_types/java/array_types/PackedAutoArrayUnionHasNoPackableFieldTest.java
package array_types.packed_auto_array_union_has_no_packable_field;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import zserio.runtime.io.SerializeUtil;

public class PackedAutoArrayUnionHasNoPackableFieldTest
{
    @Test
    public void writeRead()
    {
        final StructWithPackable array1[] = {
            new StructWithPackable("A", 65),
            new StructWithPackable("B", 66),
            new StructWithPackable("C", 67),
        };

        final StructWithPackableArray array2[] = {
            new StructWithPackableArray("ABC", new int[] { 65, 66, 67 }),
            new StructWithPackableArray("DEF", new int[] { 68, 69, 70 }),
            new StructWithPackableArray("GHI", new int[] { 71, 72, 73 }),
        };

        final UnionWithoutPackableField array3[] = new UnionWithoutPackableField[] {
            new UnionWithoutPackableField(), new UnionWithoutPackableField(), new UnionWithoutPackableField()
        };
        array3[0].setField1(4.0f);
        array3[1].setField1(1.0f);
        array3[2].setField1(0.0f);

        final PackedAutoArrayUnionHasNoPackableField packedAutoArrayUnionHasNoPackableField =
                new PackedAutoArrayUnionHasNoPackableField(array1, array2, array3);

        SerializeUtil.serializeToFile(packedAutoArrayUnionHasNoPackableField, BLOB_NAME);
        final PackedAutoArrayUnionHasNoPackableField readPackedAutoArrayUnionHasNoPackableField =
                SerializeUtil.deserializeFromFile(PackedAutoArrayUnionHasNoPackableField.class, BLOB_NAME);
        assertEquals(packedAutoArrayUnionHasNoPackableField, readPackedAutoArrayUnionHasNoPackableField);
    }

    private static final String BLOB_NAME = "packed_auto_array_union_has_no_packable_field.blob";
}

File: test/language/array_types/cpp/PackedAutoArrayUnionHasNoPackableFieldTest.cpp
#include "gtest/gtest.h"

#include "array_types/packed_auto_array_union_has_no_packable_field/PackedAutoArrayUnionHasNoPackableField.h"

#include "zserio/SerializeUtil.h"

namespace array_types
{
namespace packed_auto_array_union_has_no_packable_field
{

class PackedAutoArrayUnionHasNoPackableFieldTest : public ::testing::Test
{
protected:
    static const std::string BLOB_NAME;
};

const std::string PackedAutoArrayUnionHasNoPackableFieldTest::BLOB_NAME =
        "language/array_types/packed_auto_array_union_has_no_packable_field.blob";

TEST_F(PackedAutoArrayUnionHasNoPackableFieldTest, writeRead)
{
    PackedAutoArrayUnionHasNoPackableField packedAutoArrayUnionHasNoPackableField;

    packedAutoArrayUnionHasNoPackableField.setArray1({{
        StructWithPackable("A", 65),
        StructWithPackable("B", 66),
        StructWithPackable("C", 67)
    }});

    packedAutoArrayUnionHasNoPackableField.setArray2({{
        StructWithPackableArray("ABC", {{ 65, 66, 67 }}),
        StructWithPackableArray("DEF", {{ 68, 69, 70 }}),
        StructWithPackableArray("GHI", {{ 71, 72, 73 }})
    }});

    auto& array3 = packedAutoArrayUnionHasNoPackableField.getArray3();
    array3.resize(3);
    array3[0].setField1(4.0F);
    array3[1].setField1(1.0F);
    array3[2].setField1(0.0F);

    zserio::serializeToFile(packedAutoArrayUnionHasNoPackableField, BLOB_NAME);
    auto readPackedAutoArrayUnionHasNoPackableField =
            zserio::deserializeFromFile<PackedAutoArrayUnionHasNoPackableField>(BLOB_NAME);
    ASSERT_EQ(packedAutoArrayUnionHasNoPackableField, readPackedAutoArrayUnionHasNoPackableField);
}

} // namespace packed_auto_array_union_has_no_packable_field
} // namespace array_types
